torchgeo/trainers/instancesegmentation.py

```diff
@@ -0,0 +1,174 @@
+from typing import Any                                           
+import torch.nn as nn                                            
+import torch                                                   
+from torch import Tensor                                        
+from torchmetrics.detection.mean_ap import MeanAveragePrecision  
+from torchvision.models.detection import maskrcnn_resnet50_fpn   
+from ultralytics import YOLO  
+from .base import BaseTask  
+
+class InstanceSegmentationTask(BaseTask):
+    """
+    Task class for training and evaluating instance segmentation models.
+
+    This class supports Mask R-CNN and YOLO models and handles the following:
+    - Model configuration
+    - Loss computation
+    - Metric computation (e.g., Mean Average Precision)
+    - Training, validation, testing, and prediction steps
+    """
+
+    def __init__(
+        self,
+        model: str = 'mask_rcnn',           # Model type, e.g., 'mask_rcnn' or 'yolo'
+        backbone: str = 'resnet50',         # Backbone type for Mask R-CNN (ignored for YOLO)
+        weights: str | bool | None = None,  # Pretrained weights or custom checkpoint path
+        num_classes: int = 2,               # Number of classes, including background
+        lr: float = 1e-3,                   # Learning rate for the optimizer
+        patience: int = 10,                 # Patience for the learning rate scheduler
+        freeze_backbone: bool = False,      # Whether to freeze backbone layers (useful for transfer learning)
+    ) -> None:
+        """
+        Constructor for the InstanceSegmentationTask.
+
+        Initializes the hyperparameters, sets up the model and metrics.
+        """
+        self.weights = weights          # Save weights for model initialization
+        super().__init__()              # Initialize the BaseTask class (inherits common functionality)
+        self.save_hyperparameters()     # Save input arguments for later use (e.g., in checkpoints or logs)
+        self.model = None               # Placeholder for the model (to be initialized later)
+        self.validation_outputs = []    # List to store outputs during validation (used for debugging or analysis)
+        self.test_outputs = []          # List to store outputs during testing
+        self.configure_models()         # Call method to set up the model
+        self.configure_metrics()        # Call method to set up metrics
+
+    def configure_models(self) -> None:
+        """

```

Comment:
You should copy the docstrings from our other trainers so the documentation is uniform